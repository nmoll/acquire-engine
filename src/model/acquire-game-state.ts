export type AcquireGameState = "not started" | "started";
