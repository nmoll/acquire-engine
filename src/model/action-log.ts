import { IGameState } from "./game-state";
import { PlayerAction } from "./player-action";
import { PlayerActionResult } from "./player-action-result";

export interface ActionLog {
  action: PlayerAction;
  actionResult: PlayerActionResult;
  state: IGameState;
}
