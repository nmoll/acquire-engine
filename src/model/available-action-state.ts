import { AvailableAction } from "./available-action";

export type IAvailableActionState = AvailableAction[];
