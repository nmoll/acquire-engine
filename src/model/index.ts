export * from "./board-square-state";
export * from "./game-state";
export * from "./hotel-chain-type";
export * from "./player-turn";
export * from "./shares-state";
