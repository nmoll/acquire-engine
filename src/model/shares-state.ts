export interface ISharesState {
  [playerId: string]: {
    ["American"]?: number;
    ["Continental"]?: number;
    ["Festival"]?: number;
    ["Imperial"]?: number;
    ["Luxor"]?: number;
    ["Tower"]?: number;
    ["Worldwide"]?: number;
  };
}
