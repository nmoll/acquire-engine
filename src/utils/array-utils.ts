import { RandomUtils } from "./random-utils";

const makeNumArray = (length: number): number[] =>
  Array.from(new Array(length)).map((_, idx) => idx);

const shuffle = <T>(arr: T[], seed: number): T[] => {
  const arrCopy = [...arr];
  let len = arrCopy.length,
    elem: T,
    i: number;
  const randomGenerator = RandomUtils.randomGenerator(seed);

  while (len) {
    // Pick a remaining elementâ€¦
    i = Math.floor(randomGenerator() * len--);

    // And swap it with the current element.
    elem = arrCopy[len];
    arrCopy[len] = arrCopy[i];
    arrCopy[i] = elem;
    ++seed;
  }

  return arrCopy;
};

const unique = <T>(arr: T[]): T[] => Array.from(new Set(arr));

export const ArrayUtils = {
  shuffle,
  makeNumArray,
  unique,
};
