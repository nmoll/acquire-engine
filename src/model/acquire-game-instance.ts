import { AcquireGameState } from "./acquire-game-state";

export interface IAcquireGameInstance {
  id: string;
  randomSeed: number;
  playerIds: string[];
  hostId: string;
  state: AcquireGameState;
  isOpen?: boolean;
  createdDate?: number;
}
