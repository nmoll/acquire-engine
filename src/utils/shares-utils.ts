import { GameConfig } from "../game-config";
import { HotelChainType, ISharesState } from "../model";
import { AvailableShares } from "../model/available-shares.type";
import { HotelChainState } from "../model/hotel-chain-state";

const getAvailableSharesForHotel = (
  sharesState: ISharesState,
  hotelChain: HotelChainType
): number =>
  GameConfig.hotel.shares -
  Object.values(sharesState).reduce(
    (total, shares) => total + (shares[hotelChain] ?? 0),
    0
  );

const getAvailableShares = (
  sharesState: ISharesState,
  hotelChains: HotelChainType[]
): Partial<Record<HotelChainType, number>> =>
  hotelChains.reduce(
    (result, hotelChain) => ({
      ...result,
      [hotelChain]: getAvailableSharesForHotel(sharesState, hotelChain),
    }),
    {}
  );

const getAvailableSharesForPurchase = (
  hotelChainState: HotelChainState,
  playerCash: number
): AvailableShares =>
  Object.entries(hotelChainState).reduce<AvailableShares>(
    (result, [hotelChain, state]) => ({
      ...result,
      [hotelChain]:
        state.availableShares > 0 &&
        getSharesCost(
          hotelChain as HotelChainType,
          state.boardSquareIds.length
        ) <= playerCash,
    }),
    {}
  );

const getBasePriceBySize = (size: number): number => {
  if (size < 2) {
    // ensures a player can't buy this
    return 100000;
  }
  const priceLatter: Record<string, number> = {
    2: 0,
    3: 100,
    4: 200,
    5: 300,
    10: 400,
    20: 500,
    30: 600,
    40: 700,
    1000: 800,
  };

  const step = Object.keys(priceLatter).find(
    (threshold) => size <= Number(threshold)
  );
  return step ? priceLatter[step] : 0;
};

const getSharesCost = (
  hotelChain: HotelChainType,
  hotelSize: number
): number => {
  if (hotelSize < 2) {
    // ensures a player can't buy this
    return 100000;
  }

  return GameConfig.hotel.basePrice[hotelChain] + getBasePriceBySize(hotelSize);
};

const getMajorityBonus = (
  hotelChain: HotelChainType,
  hotelSize: number
): number => getSharesCost(hotelChain, hotelSize) * 10;

const getMinorityBonus = (
  hotelChain: HotelChainType,
  hotelSize: number
): number => getMajorityBonus(hotelChain, hotelSize) / 2;

const getMajorityAndMinorityShareholders = (
  sharesState: ISharesState,
  hotelChain: HotelChainType
): {
  majorityShareholders: string[];
  minorityShareholders: string[];
} => {
  const playersWithShares = Object.entries(sharesState)
    .filter(([, shares]) => !!shares[hotelChain])
    .map(([playerId, shares]) => ({
      playerId,
      numShares: shares[hotelChain] ?? 0,
    }));

  if (playersWithShares.length === 1) {
    return {
      majorityShareholders: [playersWithShares[0].playerId],
      minorityShareholders: [playersWithShares[0].playerId],
    };
  }

  const majorityShareAmount = Math.max(
    ...playersWithShares.map((p) => p.numShares)
  );
  const majorityShareholders = playersWithShares
    .filter((p) => p.numShares === majorityShareAmount)
    .map((p) => p.playerId);

  if (majorityShareholders.length > 1) {
    return {
      majorityShareholders,
      minorityShareholders: majorityShareholders,
    };
  }

  const minorityShareAmount = Math.max(
    ...playersWithShares
      .filter((p) => p.numShares !== majorityShareAmount)
      .map((p) => p.numShares)
  );
  const minorityShareholders = playersWithShares
    .filter((p) => p.numShares === minorityShareAmount)
    .map((p) => p.playerId);

  return {
    majorityShareholders,
    minorityShareholders,
  };
};

export const SharesUtils = {
  getAvailableShares,
  getAvailableSharesForHotel,
  getAvailableSharesForPurchase,
  getSharesCost,
  getMajorityBonus,
  getMinorityBonus,
  getMajorityAndMinorityShareholders,
};
