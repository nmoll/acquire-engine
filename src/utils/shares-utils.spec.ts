import { describe, expect, it } from "vitest";
import { ALL_HOTELS } from "../model";
import { SharesStateFactory } from "../test/factory/shares-state.factory";
import { ArrayUtils } from "./array-utils";
import { SharesUtils } from "./shares-utils";

const padNum = (num: number): string => (num < 10 ? `0${num}` : `${num}`);

describe("SharesUtils", () => {
  describe(SharesUtils.getAvailableShares.name, () => {
    it("should return 25 if no shares owned for hotel", () => {
      const sharesState = SharesStateFactory.createSharesState(`
           A C F I L T W
        P1 0 1 0 0 0 0 0
        P2 0 0 0 1 0 0 0
        P3 0 0 0 0 0 0 0
        P4 0 0 0 0 0 0 1
        `);
      expect(SharesUtils.getAvailableShares(sharesState, ["American"])).toEqual(
        {
          ["American"]: 25,
        }
      );
    });

    it("should return remaining shares if shares are owned by players", () => {
      const sharesState = SharesStateFactory.createSharesState(`
           A C F I L T W
        P1 5 1 0 0 0 3 9
        P2 1 0 0 1 0 0 9
        P3 0 0 0 0 8 7 0
        P4 2 0 0 0 1 2 1
        `);

      const hotels = ALL_HOTELS.map((hotel) => hotel);

      expect(SharesUtils.getAvailableShares(sharesState, hotels)).toEqual({
        ["American"]: 17,
        ["Continental"]: 24,
        ["Festival"]: 25,
        ["Imperial"]: 24,
        ["Luxor"]: 16,
        ["Tower"]: 13,
        ["Worldwide"]: 6,
      });
    });
  });

  describe(SharesUtils.getSharesCost.name, () => {
    it("should calculate share price for the given hotel chain and size", () => {
      const hotelSizes = ArrayUtils.makeNumArray(45);
      const hotelPricesBySize = ALL_HOTELS.reduce(
        (accHotels, hotel) => ({
          ...accHotels,
          [hotel]: hotelSizes.reduce(
            (accSizes, hotelSize) => ({
              ...accSizes,
              [padNum(hotelSize)]: SharesUtils.getSharesCost(hotel, hotelSize),
            }),
            {}
          ),
        }),
        {}
      );

      expect(hotelPricesBySize).toMatchSnapshot();
    });
  });

  describe(SharesUtils.getMajorityBonus.name, () => {
    it("should calculate majority bonus for the given hotel chain and size", () => {
      const hotelSizes = ArrayUtils.makeNumArray(45);
      const hotelPricesBySize = ALL_HOTELS.reduce(
        (accHotels, hotel) => ({
          ...accHotels,
          [hotel]: hotelSizes.reduce(
            (accSizes, hotelSize) => ({
              ...accSizes,
              [padNum(hotelSize)]: SharesUtils.getMajorityBonus(
                hotel,
                hotelSize
              ),
            }),
            {}
          ),
        }),
        {}
      );

      expect(hotelPricesBySize).toMatchSnapshot();
    });
  });

  describe(SharesUtils.getMinorityBonus.name, () => {
    it("should calculate minority bonus for the given hotel chain and size", () => {
      const hotelSizes = ArrayUtils.makeNumArray(45);
      const hotelPricesBySize = ALL_HOTELS.reduce(
        (accHotels, hotel) => ({
          ...accHotels,
          [hotel]: hotelSizes.reduce(
            (accSizes, hotelSize) => ({
              ...accSizes,
              [padNum(hotelSize)]: SharesUtils.getMinorityBonus(
                hotel,
                hotelSize
              ),
            }),
            {}
          ),
        }),
        {}
      );

      expect(hotelPricesBySize).toMatchSnapshot();
    });
  });

  describe(SharesUtils.getMajorityAndMinorityShareholders.name, () => {
    it("should return empty lists if no players own any shares", () => {
      const sharesState = SharesStateFactory.createSharesState(`
           A C F I L T W
        P1 0 0 0 0 2 0 0
        P2 0 1 0 1 0 3 0
        P3 0 0 1 0 0 0 0
        P4 0 0 0 0 0 0 5
        `);

      expect(
        SharesUtils.getMajorityAndMinorityShareholders(sharesState, "American")
      ).toEqual({
        majorityShareholders: [],
        minorityShareholders: [],
      });
    });

    it("should return single player for majority and minority if no one else has shares", () => {
      const sharesState = SharesStateFactory.createSharesState(`
             A C F I L T W
          P1 0 4 0 0 0 0 0
          P2 0 0 0 1 0 0 0
          P3 0 0 0 0 0 0 0
          P4 0 0 0 0 0 0 1
          `);

      expect(
        SharesUtils.getMajorityAndMinorityShareholders(
          sharesState,
          "Continental"
        )
      ).toEqual({
        majorityShareholders: ["1"],
        minorityShareholders: ["1"],
      });
    });

    it("should return multiple players for majority and minority if they tie for majority shares", () => {
      const sharesState = SharesStateFactory.createSharesState(`
             A C F I L T W
          P1 0 4 0 3 0 0 0
          P2 0 0 0 6 0 0 0
          P3 0 0 0 1 0 0 0
          P4 0 0 0 6 0 0 1
          `);

      expect(
        SharesUtils.getMajorityAndMinorityShareholders(sharesState, "Imperial")
      ).toEqual({
        majorityShareholders: ["2", "4"],
        minorityShareholders: ["2", "4"],
      });
    });

    it("should return multiple players for minority if they tie for minority shares", () => {
      const sharesState = SharesStateFactory.createSharesState(`
             A C F I L T W
          P1 1 4 0 3 0 0 0
          P2 1 0 0 6 0 0 0
          P3 1 0 0 1 0 0 0
          P4 2 0 0 6 0 0 1
          `);

      expect(
        SharesUtils.getMajorityAndMinorityShareholders(sharesState, "American")
      ).toEqual({
        majorityShareholders: ["4"],
        minorityShareholders: ["1", "2", "3"],
      });
    });

    it("should return single majority and minority share holder", () => {
      const sharesState = SharesStateFactory.createSharesState(`
             A C F I L T W
          P1 1 4 0 3 0 0 0
          P2 1 0 0 6 0 0 0
          P3 1 0 0 1 0 0 0
          P4 2 0 0 0 0 0 1
          `);

      expect(
        SharesUtils.getMajorityAndMinorityShareholders(sharesState, "Imperial")
      ).toEqual({
        majorityShareholders: ["2"],
        minorityShareholders: ["1"],
      });
    });
  });

  describe(SharesUtils.getNextPlayerWithOrphanedShares.name, () => {
    it("should return the next player with shares of the given hotel", () => {
      const sharesState = SharesStateFactory.createSharesState(`
           A C F I L T W
        P1 1 0 0 0 2 0 0
        P2 0 1 0 1 0 3 0
        P3 5 0 1 0 0 0 0
        P4 0 0 0 0 0 0 5
        `);

      expect(
        SharesUtils.getNextPlayerWithOrphanedShares(
          sharesState,
          "1",
          "American",
          []
        )
      ).toEqual({ playerId: "1", remainingShares: 1 });

      expect(
        SharesUtils.getNextPlayerWithOrphanedShares(
          sharesState,
          "2",
          "American",
          []
        )
      ).toEqual({ playerId: "3", remainingShares: 5 });

      expect(
        SharesUtils.getNextPlayerWithOrphanedShares(
          sharesState,
          "3",
          "American",
          []
        )
      ).toEqual({ playerId: "3", remainingShares: 5 });

      expect(
        SharesUtils.getNextPlayerWithOrphanedShares(
          sharesState,
          "4",
          "American",
          []
        )
      ).toEqual({ playerId: "1", remainingShares: 1 });
    });

    it("should return null if no player owns shares for the given hotel", () => {
      const sharesState = SharesStateFactory.createSharesState(`
           A C F I L T W
        P1 1 0 0 0 2 0 0
        P2 0 1 0 1 0 3 0
        P3 5 0 0 0 0 0 0
        P4 0 0 0 0 0 0 5
        `);

      expect(
        SharesUtils.getNextPlayerWithOrphanedShares(
          sharesState,
          "1",
          "Festival",
          []
        )
      ).toEqual(null);
    });
  });
});
