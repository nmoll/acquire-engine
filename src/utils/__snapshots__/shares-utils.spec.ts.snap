// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SharesUtils > getSharesCost > should calculate share price for the given hotel chain and size 1`] = `
{
  "American": {
    "00": 100000,
    "01": 100000,
    "02": 300,
    "03": 400,
    "04": 500,
    "05": 600,
    "06": 700,
    "07": 700,
    "08": 700,
    "09": 700,
    "10": 700,
    "11": 800,
    "12": 800,
    "13": 800,
    "14": 800,
    "15": 800,
    "16": 800,
    "17": 800,
    "18": 800,
    "19": 800,
    "20": 800,
    "21": 900,
    "22": 900,
    "23": 900,
    "24": 900,
    "25": 900,
    "26": 900,
    "27": 900,
    "28": 900,
    "29": 900,
    "30": 900,
    "31": 1000,
    "32": 1000,
    "33": 1000,
    "34": 1000,
    "35": 1000,
    "36": 1000,
    "37": 1000,
    "38": 1000,
    "39": 1000,
    "40": 1000,
    "41": 1100,
    "42": 1100,
    "43": 1100,
    "44": 1100,
  },
  "Continental": {
    "00": 100000,
    "01": 100000,
    "02": 400,
    "03": 500,
    "04": 600,
    "05": 700,
    "06": 800,
    "07": 800,
    "08": 800,
    "09": 800,
    "10": 800,
    "11": 900,
    "12": 900,
    "13": 900,
    "14": 900,
    "15": 900,
    "16": 900,
    "17": 900,
    "18": 900,
    "19": 900,
    "20": 900,
    "21": 1000,
    "22": 1000,
    "23": 1000,
    "24": 1000,
    "25": 1000,
    "26": 1000,
    "27": 1000,
    "28": 1000,
    "29": 1000,
    "30": 1000,
    "31": 1100,
    "32": 1100,
    "33": 1100,
    "34": 1100,
    "35": 1100,
    "36": 1100,
    "37": 1100,
    "38": 1100,
    "39": 1100,
    "40": 1100,
    "41": 1200,
    "42": 1200,
    "43": 1200,
    "44": 1200,
  },
  "Festival": {
    "00": 100000,
    "01": 100000,
    "02": 300,
    "03": 400,
    "04": 500,
    "05": 600,
    "06": 700,
    "07": 700,
    "08": 700,
    "09": 700,
    "10": 700,
    "11": 800,
    "12": 800,
    "13": 800,
    "14": 800,
    "15": 800,
    "16": 800,
    "17": 800,
    "18": 800,
    "19": 800,
    "20": 800,
    "21": 900,
    "22": 900,
    "23": 900,
    "24": 900,
    "25": 900,
    "26": 900,
    "27": 900,
    "28": 900,
    "29": 900,
    "30": 900,
    "31": 1000,
    "32": 1000,
    "33": 1000,
    "34": 1000,
    "35": 1000,
    "36": 1000,
    "37": 1000,
    "38": 1000,
    "39": 1000,
    "40": 1000,
    "41": 1100,
    "42": 1100,
    "43": 1100,
    "44": 1100,
  },
  "Imperial": {
    "00": 100000,
    "01": 100000,
    "02": 400,
    "03": 500,
    "04": 600,
    "05": 700,
    "06": 800,
    "07": 800,
    "08": 800,
    "09": 800,
    "10": 800,
    "11": 900,
    "12": 900,
    "13": 900,
    "14": 900,
    "15": 900,
    "16": 900,
    "17": 900,
    "18": 900,
    "19": 900,
    "20": 900,
    "21": 1000,
    "22": 1000,
    "23": 1000,
    "24": 1000,
    "25": 1000,
    "26": 1000,
    "27": 1000,
    "28": 1000,
    "29": 1000,
    "30": 1000,
    "31": 1100,
    "32": 1100,
    "33": 1100,
    "34": 1100,
    "35": 1100,
    "36": 1100,
    "37": 1100,
    "38": 1100,
    "39": 1100,
    "40": 1100,
    "41": 1200,
    "42": 1200,
    "43": 1200,
    "44": 1200,
  },
  "Luxor": {
    "00": 100000,
    "01": 100000,
    "02": 200,
    "03": 300,
    "04": 400,
    "05": 500,
    "06": 600,
    "07": 600,
    "08": 600,
    "09": 600,
    "10": 600,
    "11": 700,
    "12": 700,
    "13": 700,
    "14": 700,
    "15": 700,
    "16": 700,
    "17": 700,
    "18": 700,
    "19": 700,
    "20": 700,
    "21": 800,
    "22": 800,
    "23": 800,
    "24": 800,
    "25": 800,
    "26": 800,
    "27": 800,
    "28": 800,
    "29": 800,
    "30": 800,
    "31": 900,
    "32": 900,
    "33": 900,
    "34": 900,
    "35": 900,
    "36": 900,
    "37": 900,
    "38": 900,
    "39": 900,
    "40": 900,
    "41": 1000,
    "42": 1000,
    "43": 1000,
    "44": 1000,
  },
  "Tower": {
    "00": 100000,
    "01": 100000,
    "02": 200,
    "03": 300,
    "04": 400,
    "05": 500,
    "06": 600,
    "07": 600,
    "08": 600,
    "09": 600,
    "10": 600,
    "11": 700,
    "12": 700,
    "13": 700,
    "14": 700,
    "15": 700,
    "16": 700,
    "17": 700,
    "18": 700,
    "19": 700,
    "20": 700,
    "21": 800,
    "22": 800,
    "23": 800,
    "24": 800,
    "25": 800,
    "26": 800,
    "27": 800,
    "28": 800,
    "29": 800,
    "30": 800,
    "31": 900,
    "32": 900,
    "33": 900,
    "34": 900,
    "35": 900,
    "36": 900,
    "37": 900,
    "38": 900,
    "39": 900,
    "40": 900,
    "41": 1000,
    "42": 1000,
    "43": 1000,
    "44": 1000,
  },
  "Worldwide": {
    "00": 100000,
    "01": 100000,
    "02": 300,
    "03": 400,
    "04": 500,
    "05": 600,
    "06": 700,
    "07": 700,
    "08": 700,
    "09": 700,
    "10": 700,
    "11": 800,
    "12": 800,
    "13": 800,
    "14": 800,
    "15": 800,
    "16": 800,
    "17": 800,
    "18": 800,
    "19": 800,
    "20": 800,
    "21": 900,
    "22": 900,
    "23": 900,
    "24": 900,
    "25": 900,
    "26": 900,
    "27": 900,
    "28": 900,
    "29": 900,
    "30": 900,
    "31": 1000,
    "32": 1000,
    "33": 1000,
    "34": 1000,
    "35": 1000,
    "36": 1000,
    "37": 1000,
    "38": 1000,
    "39": 1000,
    "40": 1000,
    "41": 1100,
    "42": 1100,
    "43": 1100,
    "44": 1100,
  },
}
`;
